File: site/tools/generate-docs/message.ts
import Field from './field';
import { Package } from './package';
import { JsonMessage } from './types';

const { log } = console;

export class Message {
  package: Package;
  fileName: string;

  name: string;
  longName: string;
  fullName: string;
  description: string;
  fields: Field[];

  get hasFields() {
    return this.fields.length > 0;
  }

  constructor(json: JsonMessage, pkg: Package, fileName: string) {
    log(`Creating message ${json.name} with ${json.fields.length} fields`);
    this.package = pkg;
    this.fileName = fileName;
    this.name = json.name;
    this.longName = json.longName;
    this.fullName = json.fullName;
    this.description = json.description;
    this.fields = json.fields.map((f) => new Field(f, this));
  }
}

File: models/message.go
package models

import (
	"fmt"

	gendoc "github.com/pseudomuto/protoc-gen-doc"
)

type Message struct {
	Package  *Package
	FileName string

	Name        string
	LongName    string
	FullName    string
	Description string
	Fields      []*Field

	// Source is the github url to the exact line in the proto file where
	// this message is declared. This field is set in Method.Request and
	// Method.Response
	Source string
}

// NewMessage creates a new Message model from a Message definition
func NewMessage(messageDef *gendoc.Message, pkg *Package,
	fileName string) *Message {

	fmt.Printf("Creating message %s with %d fields\n", messageDef.Name,
		len(messageDef.Fields))

	m := &Message{
		Package:     pkg,
		FileName:    fileName,
		Name:        messageDef.Name,
		LongName:    messageDef.LongName,
		FullName:    messageDef.FullName,
		Description: messageDef.Description,

		Fields: make([]*Field, 0),
	}
	for _, fieldDef := range messageDef.Fields {
		field := NewField(*fieldDef, m)
		m.Fields = append(m.Fields, field)
	}

	return m
}

// HasFields returns true if the message has fields
func (m *Message) HasFields() bool {
	return len(m.Fields) > 0
}
